diffkemp/simpll/DifferentialFunctionComparator.cpp

```diff
@@ -1845,3 +1895,49 @@ bool DifferentialFunctionComparator::isDependingOnReloc(
 
     return false;
 }
+
+/// Recursively collect operands of commutative binary operators.
+/// Leafs must be constants or already synchronized values.
+/// Return false if a non-synchronized non-constant leaf is found.
+bool DifferentialFunctionComparator::collectBinaryOperands(
+        const Value *Val,
+        Instruction::BinaryOps Opcode,
+        std::multiset<int> &SNs,
+        std::multiset<std::string> &Constants,
+        DenseMap<const Value *, int> &sn_map) const {
+    // Substitute an ignored value with its replacement when necessary
+    auto replacementIt = ignoredWithReplacement.find(Val);
+    const Value *newVal = replacementIt == ignoredWithReplacement.end()
+                                  ? Val
+                                  : replacementIt->second;
+    // Constant leaf
+    if (auto Const = dyn_cast<ConstantInt>(newVal)) {
+        Constants.insert(valueAsString(Const));
```

Comment:
Why not use the value directly?

Reply:
Ah, good point. I think I wanted to avoid storing raw integer values because I thought a similar issue might occur as the one solved in https://github.com/viktormalik/diffkemp/pull/301/commits/dfccede263c2892f5507557373120374cf5a6eba (i.e., the constant being too large to fit into 64 bits). I did not realize that I could use the intermediate representation `llvm::APInt`.

I ran my experiments again, and the mentioned issue does not occur there (but I still think it could theoretically occur).

So there are three options:
1. Use `int64_t` and pray that such large integers do not get used in arithmetic operations.

2. Use `llvm::APInt`. However, `llvm:APInt`  does not implement operator `<` (because it has both signed and unsigned comparison functions) and thus cannot be used in `std::set` without creating a derived class or more obscure C++ magic (e.g., functors). 

3. Leave it as it is.

Reply:
Could we use 2. with `std::unordered_multiset`? I don't expect the sets to become too large for this to be a problem. Otherwise let's just stick with 1. for simplicity as the pattern should be replaced by a more complex solution anyways.

Reply:
Sticking with`int64_t` since `llvm::APInt` cannot be hashed by default.